File: Inboxes/CSharp/Cli/InboxManager.cs
namespace InboxesCli;

public class Message
{
    public int Id { get; set; }
    public string From { get; set; } = "";
    public string Subject { get; set; } = "";
    public string Body { get; set; } = "";
    public bool Read { get; set; }
    public string CreatedAt { get; set; } = "";
}

public class InboxManager
{
    private readonly List<Message> _messages = [];
    private int _nextId = 1;

    public Message Send(string from, string subject, string body)
    {
        var msg = new Message
        {
            Id = _nextId++,
            From = from,
            Subject = subject,
            Body = body,
            Read = false,
            CreatedAt = DateTime.UtcNow.ToString("o"),
        };
        _messages.Add(msg);
        return msg;
    }

    public List<Message> List()
    {
        return [.. _messages];
    }

    public Message? Read(int id)
    {
        var msg = _messages.Find(m => m.Id == id);
        if (msg != null) msg.Read = true;
        return msg;
    }

    public bool Delete(int id)
    {
        return _messages.RemoveAll(m => m.Id == id) > 0;
    }
}

File: Inboxes/Java/Cli/src/main/java/com/portfolio/inboxes/InboxManager.java
package com.portfolio.inboxes;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class InboxManager {
    private final List<Message> messages = new ArrayList<>();
    private int nextId = 1;

    public List<Message> list() {
        return new ArrayList<>(messages);
    }

    public Map<String, String> send(String from, String subject, String body) {
        Map<String, String> errors = MessageValidator.validate(from, subject, body);
        if (!errors.isEmpty()) {
            return errors;
        }
        messages.add(new Message(nextId++, from.trim(), subject.trim(), body.trim()));
        return errors;
    }

    public Message read(int id) {
        for (Message message : messages) {
            if (message.getId() == id) {
                message.setRead(true);
                return message;
            }
        }
        return null;
    }

    public boolean delete(int id) {
        for (int i = 0; i < messages.size(); i++) {
            if (messages.get(i).getId() == id) {
                messages.remove(i);
                return true;
            }
        }
        return false;
    }
}
